### ISSUE TITLE
[Foundry Ingress] ERROR c.a.daemon.serial.IngressThread - I/O error.

### ISSUE THREAD

**Issue by junming** [abstractfoundry/lumicube-daemon#13]
Hello,

I just received the Lumicube today, very nice piece of hardware!
However I just had a couple issues.
I'm using Pi OS most recent version, I just got it and flashed it to a SD card today.
I solved the first one:

```
pi@raspberrypi:~/AbstractFoundry/Daemon $ ./launch.sh 
Exception in thread "main" java.lang.IllegalStateException: I/O exception whilst installing the virtual microphone.
	at com.abstractfoundry.daemon.audio.VirtualMicrophone.<init>(VirtualMicrophone.java:53)
	at com.abstractfoundry.daemon.Daemon.<init>(Daemon.java:69)
	at com.abstractfoundry.daemon.Daemon.main(Daemon.java:113)
Caused by: java.io.IOException: Cannot run program "/usr/bin/pactl": error=2, No such file or directory
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1142)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
	at com.abstractfoundry.daemon.audio.VirtualMicrophone.getPulseAudioSymbolicNames(VirtualMicrophone.java:107)
	at com.abstractfoundry.daemon.audio.VirtualMicrophone.<init>(VirtualMicrophone.java:31)
	... 2 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:313)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:244)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1109)
	... 5 more
```

I solved it by installing PulseAudio (documenting it in case it helps anyone), I think it was supposed to be installed by the install.py script but somehow it didn't happen:
```
sudo apt install pulseaudio pavucontrol paprefs
```

Then the second issue which I haven't been able to solve yet, here is an example of the stacktrace:

```
22:20:45.707 [Foundry Ingress] ERROR c.a.daemon.serial.IngressThread - I/O error.
com.fazecast.jSerialComm.SerialPortIOException: This port appears to have been shutdown or disconnected.
	at com.fazecast.jSerialComm.SerialPort$SerialPortInputStream.read(SerialPort.java:1494)
	at com.abstractfoundry.daemon.serial.IngressThread.run(IngressThread.java:51)
22:20:46.051 [Foundry Outbox Disruptor] WARN  c.a.daemon.serial.SerialDriver - I/O error whilst transmitting frame.
com.fazecast.jSerialComm.SerialPortIOException: This port appears to have been shutdown or disconnected.
	at com.fazecast.jSerialComm.SerialPort$SerialPortOutputStream.write(SerialPort.java:1591)
	at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:81)
	at java.base/java.io.BufferedOutputStream.write(BufferedOutputStream.java:127)
	at com.abstractfoundry.daemon.serial.SerialDriver.handleOutbox(SerialDriver.java:185)
	at com.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:168)
	at com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
	at java.base/java.lang.Thread.run(Thread.java:831)
22:20:46.709 [Foundry Ingress] ERROR c.a.daemon.serial.IngressThread - I/O error.
com.fazecast.jSerialComm.SerialPortIOException: This port appears to have been shutdown or disconnected.
	at com.fazecast.jSerialComm.SerialPort$SerialPortInputStream.read(SerialPort.java:1494)
	at com.abstractfoundry.daemon.serial.IngressThread.run(IngressThread.java:51)
```

I know it seems to be linked to jSerialComm, but it is also part of the daemon, and while i can debug python, compiled java is beyond my expertise... also wondering if anyone else is having the same problem?
As you may guess, the Daemon runs, but is not usable.

**Comment by mfoo** [comment]
With the lumicube set up, run the following:

```
$ ls -l /dev/ttyS*
crw--w---- 1 root tty 4, 64 Sep 29 06:45 /dev/ttyS0
```

From the output above you can see that the serial device is owned by `root` and group-owned by `tty`. You might also see `dialout` instead of `tty` here, it depends on the OS.

That means that the user you're running the process with needs to be in that group so that it can talk to the serial port that's used for controlling the LEDs.

You can run `groups` to see what groups your user is in, here this user is not in the `tty` group:

```
ubuntu@k8s-master-1:~$ groups
ubuntu adm dialout cdrom floppy sudo audio dip video plugdev netdev lxd
```

If your user isn't in that group, you can then add your user to the group like this:

```
sudo usermod -a -G tty youruser
```

(swapping `tty` for the group you see from the first command and `youruser` for your user name)

After that reboot the pi and let us know if you get further :)

If your user was in the group then it looks like it's a different problem, and in that case it'd be useful to see the output of the following:

```
$ cat /boot/**/cmdline.txt
$ grep -r uart /boot/**/*.txt
```

**Comment by abstractfoundry** [comment]
Hi @junming, and thanks @mfoo for helping fix this! I'm assuming you're not using the 32-bit desktop version of the official Raspberry Pi OS, but the "lite" version. We only support the desktop version of the OS at the moment, so that's why some of the dependencies we 're expecting to be present are missing from your installation. Did @mfoo's suggestion fix your error, or are you still having issues? If you are still experiencing problems, we would recommend using the 32-bit desktop version of the official Raspberry Pi OS instead.

**Comment by junming** [comment]
Thanks both @mfoo for your help and @abstractfoundry for getting back to me!
I did install the Lite version at first, as I didn't expect there was a need for the Desktop version.
In the day following my question, I noticed there was a discord server, which I joined, and there I was advised to try with the desktop version. I did and everything worked perfectly out of the box.
I am sure @mfoo's workaround works as well as I noticed it was also mentioned by a few enthusiasts on discord.
Thanks for the great project! I made some python scripts myself and after I take some time to clean them I will share them with the community!
